### PR TITLE
Add unit tests for snow provider and apiBuilder

### PR DESCRIPTION
**Pull request by jiayiwang7**
*Issue #, if available:*

#1105 

*Description of changes:*

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/snow/snow_test.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	tt := newSnowTest(t)
+	tt.Expect(tt.provider.Name()).To(Equal(expectedSnowProviderName))
+}
+
+func TestSetupAndValidateCreateClusterSuccess(t *testing.T) {
+	tt := newSnowTest(t)
+	setupContext(t, credsFileEnvVar)
+	setupContext(t, certsFileEnvVar)
+	os.Setenv(credsFileEnvVar, credsFilePath)
+	os.Setenv(certsFileEnvVar, certsFilePath)
```
Comment (g-gaston) [inline pkg/providers/snow/snow_test.go:125]: Do we need to call these two methods twice?

File: pkg/providers/snow/snow_test.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	err := tt.provider.SetupAndValidateCreateCluster(tt.ctx, tt.clusterSpec)
+	tt.Expect(err).To(Succeed())
+}
+
+func TestSetupAndValidateCreateClusterNoCredsEnv(t *testing.T) {
+	tt := newSnowTest(t)
+	setupContext(t, credsFileEnvVar)
+	setupContext(t, certsFileEnvVar)
+	os.Setenv(certsFileEnvVar, certsFilePath)
```
Comment (g-gaston) [inline pkg/providers/snow/snow_test.go:134]: Should this be called inside `setupContext`?

File: pkg/providers/snow/snow_test.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+func TestSetupAndValidateCreateClusterNoCertsEnv(t *testing.T) {
+	tt := newSnowTest(t)
+	setupContext(t, credsFileEnvVar)
+	os.Setenv(credsFileEnvVar, credsFilePath)
+	err := tt.provider.SetupAndValidateCreateCluster(tt.ctx, tt.clusterSpec)
+	tt.Expect(err).To(MatchError(ContainSubstring("EKSA_SNOW_DEVICES_CA_BUNDLES_FILE is not set or is empty")))
+}
+
+// TODO: add more tests (multi worker node groups, unstacked etcd, etc.)
```
Comment (g-gaston) [inline pkg/providers/snow/snow_test.go:147]: Suggestion: maybe keep this method tests at a minimum and just test the methods that return api structs? That avoids having to create a ton of yaml files and probably makes tests more "unitary"

File: pkg/providers/snow/snow_test.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+type apiObjects struct {
+	cluster          *v1alpha1.Cluster
+	clusterSpec      *cluster.Spec
+	datacenterConfig *v1alpha1.SnowDatacenterConfig
+	machineConfigs   map[string]*v1alpha1.SnowMachineConfig
+}
+
+func givenFullAPIObjects(t *testing.T, fileName string) apiObjects {
+	clusterSpec := test.NewFullClusterSpec(t, path.Join(testDataDir, fileName))
```
Comment (g-gaston) [inline pkg/providers/snow/snow_test.go:65]: Can we build this without using a yaml file? Just defining structs in code
I know that vsphere does it this way but I think structs will be more flexible

**Review by g-gaston — COMMENTED** [review]
I would add some tests for the snow apibuilder methods in the same way you did for the generic clusterapi ones

File: pkg/providers/snow/snow_test.go (hunk 3 above)
Comment (jiayiwang7) [reply pkg/providers/snow/snow_test.go:147]: not sure if i 100% agree. this method returns cp and md yaml templates in []byte. it makes sense to test against yamls. 
also the api struct test is covered in the two apibuilder_test

**Comment by jiayiwang7** [comment]
> I would add some tests for the snow apibuilder methods in the same way you did for the generic clusterapi ones

added.

**Review by g-gaston — APPROVED** [review]
/lgtm

**Pull request merged**